File: yukino/teams/registry.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package teams

import "sync"

// NameRegistry is the process-wide global name registry, maintaining
// member name → agentID mappings. SendMessage uses it to resolve the recipient
// name given by the user/LLM into a delivery identifier.
type NameRegistry struct {
	mu    sync.Mutex
	names map[string]string // name -> agentID
}

var (
	nameRegistryOnce     sync.Once
	nameRegistryInstance *NameRegistry
)

// GetNameRegistry returns the global singleton registry.
func GetNameRegistry() *NameRegistry {
	nameRegistryOnce.Do(func() {
		nameRegistryInstance = &NameRegistry{names: make(map[string]string)}
	})
	return nameRegistryInstance
}

// Register records a name → agentID mapping.
func (r *NameRegistry) Register(name, agentID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.names[name] = agentID
}

// Resolve resolves a name or ID into an agentID: it looks up the name first;
// if the input is itself an ID, it is returned as-is; otherwise an empty
// string is returned.
func (r *NameRegistry) Resolve(nameOrID string) string {
	r.mu.Lock()
	defer r.mu.Unlock()
	if id, ok := r.names[nameOrID]; ok {
		return id
	}
	for _, id := range r.names {
		if id == nameOrID {
			return nameOrID
		}
	}
	return ""
}

// Unregister removes the mapping for a name.
func (r *NameRegistry) Unregister(name string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.names, name)
}

// Clear removes all mappings; mainly used for test isolation.
func (r *NameRegistry) Clear() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.names = make(map[string]string)
}

File: apps/yukino/src/teams/registry.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * In-process global name registry maintaining a member-name -> agentId mapping.
 * SendMessage uses it to resolve recipient names provided by the user/LLM into delivery identifiers.
 */
export class NameRegistry {
  private names = new Map<string, string>();

  /** Registers a name -> agentId mapping. */
  register(name: string, agentId: string): void {
    this.names.set(name, agentId);
  }

  /**
   * Resolves a name or ID to an agentId: looks up by name first; if the input is already an ID, returns it as-is;
   * returns undefined if neither lookup succeeds.
   */
  resolve(nameOrId: string): string | undefined {
    const byName = this.names.get(nameOrId);
    if (byName !== undefined) {
      return byName;
    }
    for (const id of this.names.values()) {
      if (id === nameOrId) {
        return nameOrId;
      }
    }
    return undefined;
  }

  /** Removes a name mapping. */
  unregister(name: string): void {
    this.names.delete(name);
  }

  /** Clears all mappings; primarily used for test isolation. */
  clear(): void {
    this.names.clear();
  }
}

let instance: NameRegistry | undefined;

/** Returns the global singleton registry. */
export function getNameRegistry(): NameRegistry {
  instance ??= new NameRegistry();
  return instance;
}
